### DISCUSSION TITLE
Access request headers inside custom logger

### DISCUSSION THREAD

**Discussion by hrishikesh-k** [fastify/fastify#5001] [Q&A] [answered]
Hello!

I wish to pass a custom logger to Fastify (majorly to customise or have full control over how logs are presented). This is what I currently have:

```ts
export class Logger {
  level : string
  constructor(level : string = '30') {
    this.level = level
  }
  child() {
    return new Logger()
  }
  debug(message) {
    // here I wish to access request headers
  }
  error() {}
  fatal() {}
  info(message) {
    console.log('info:', message)
  }
  silent() (}
  trace() {}
  warn() {}
}
```

Inside my `debug` method, I wish to write something like:

```ts
if (request.headers['x-nf-api-debug-logging']?.toString() === '1') {
  console.log(message}
}
```

This way, I could simply pass a debug header to my API whenever I wish to see some extra logging. I know that, in my request, I could possibly pass an extra param like:

```ts
request.log.debug('debug logging', request.headers['x-nf-api-debug-logging'])
```

and compare that in my method. However, I also need to be able to use debug logging outside of the request (for example, when signing and unsigning the cookie) and I cannot access the request headers there.

When initializing Fastify, I've passed the `disableRequestLogging: true` option as I can handle logging the request and reply logs myself.

I'll be deploying this as an API for my application on Netlify Functions (using AWS Lambda under-the-hood), in case that makes a 
difference. An alternative I could take is to use environment variables, but those have 2 limitations:

1. It will be an all-or-none switch. All requests that would hit my API till I have this variable on would have debug logging turned on, or all would have it off. I can't debug it myself on a per-request basis.
2. Environment Variables on Netlify won't be updated in my API until I build and deploy my application again. So whenever I need to change the value of the variable, I'd have to build and deploy the entire thing again. It takes less than a minute, but it's some extra work I'd like to avoid if possible.

Is there a way to achieve this?

**Comment by metcoder95** [comment]
You can always pass a custom instance of `pino` through the `logger` option when creating a fastify app. e.g. `fastify({ logger: pino(opts) })`.

Your use case seems to work perfectly with either [`serializers`](https://getpino.io/#/docs/api?id=serializers-object) or [`mixins`](https://getpino.io/#/docs/api?id=mixin-function) from pino. Have you considered it already?

**Reply by hrishikesh-k** [reply]
Thank you for your response. I had tried that before going down this path, but sadly couldn't find a way. I gave it a shot again, and even after reading the docs multiple times, I don't see how I can achieve what I need using those options. So far, I have configured this:

```ts
disableRequestLogging: true,
logger: {
  base: {},
  level: 'debug',
  timestamp: false
},
requestIdHeader: 'x-nf-request-id'
```

which gives me logs like (when using `` request.log.info(`responding with ${reply.statusCode}`) ``):

```JSON
{"level":30,"reqId":"01H9058P030JAK4BPWV8JCEJE0","msg":"responding with 200"}
```

Ideally, I would like to format it, and unfortunately I cannot use pino-pretty because it needs some special kind of bundling: https://getpino.io/#/docs/bundling. However, prettifying logs is secondary if I can get my request header thing to work with pino.

I tried adding a basic serailizer like:

```ts
msg: (foo, bar) => {
  console.log(bar)
  return foo
}
```

As I can understand `foo` is the data that's being passed to `msg` (basically my actual log data) and `bar` is `undefined` (so there's no second argument). I don't see how I can use this to access request headers. Same goes for mixins. By reading the documentation, I don't see a way to access the request headers in there.

Do you mind sharing a rough example of how this can be done?

**Reply by hrishikesh-k** [reply]
I got `pino-pretty` working, so that's one problem solved.

**Comment by hrishikesh-k** [comment, marked as answer]
Finally, found a way to achieve what I wanted! As mentioned, the only reason I wanted to access request headers inside custom logger was to dynamically change log level based on request headers. So here's my new Logger class:

```ts
const logLevels = [
  'debug',
  'error',
  'fatal',
  'info',
  'silent',
  'trace',
  'warn'
] as const
type LogLevel = typeof logLevels[number]
export class Logger {
  #_level : LogLevel
  #levelNumber : number
  #levelToNumber : Record<LogLevel, number> = {
    debug: 20,
    error: 50,
    info: 30,
    fatal: 60,
    trace: 10,
    silent: 0,
    warn: 40
  }
  child() {
    return new Logger()
  }
  constructor(level : LogLevel = 'info') {
    this.#_level = level
    this.#levelNumber = this.#levelToNumber[this.#_level]
  }
  debug(msg : string) {
    if (this.#levelNumber && this.#levelNumber <= this.#levelToNumber.debug) {
      console.debug(msg)
    }
  }
  error(msg : string) {
    if (this.#levelNumber && this.#levelNumber <= this.#levelToNumber.error) {
      console.error(msg)
    }
  }
  get level() {
    return this.#_level
  }
  fatal(msg : string) {
    if (this.#levelNumber && this.#levelNumber <= this.#levelToNumber.fatal) {
      console.error(msg)
    }
  }
  info(msg : string) {
    if (this.#levelNumber && this.#levelNumber <= this.#levelToNumber.info) {
      console.info(msg)
    }
  }
  set level(newLevel) {
    this.#_level = newLevel
    this.#levelNumber = this.#levelToNumber[this.#_level]
  }
  silent() {}
  trace(msg : string) {
    if (this.#levelNumber && this.#levelNumber <= this.#levelToNumber.trace) {
      console.trace(msg)
    }
  }
  warn(msg : string) {
    if (this.#levelNumber && this.#levelNumber <= this.#levelToNumber.warn) {
      console.warn(msg)
    }
  }
}
```

and in my Fastify-side of things, I have:

```ts
fastify.addHook('onRequest', (req, _res, done) => {
  const logLevelHeader = req.headers['x-log-level']
  if (logLevels.includes(logLevelHeader)) {
    req.log.level = logLevelHeader
  }
  done()
})
```

Now, simply making a request to my API with a header like `x-log-level: debug` sets the log level to debug.
